### PR TITLE
refactoring of KORE + linking backend on KORE + tiny backend

### PR DESCRIPTION
**Pull request by cos**
**please review**

This pull request bundles three major modifications:
- extracting inner KORE interfaces and further simplification of the data structures
- making the Java backend accept KORE input -- Andrei's contribution
- tiny backend -- my experimental backend; it is put in a separate module and no module except k-distribution depends on it. This means code can only depend on it through interfaces, so the Java backend, when ready, can substitute it without change.

### CODE REVIEW

**Comment by cos** [comment]
Jenkins, test this please.

**Comment by cos** [comment]
Jenkins, test this please.

**Comment by cos** [comment]
@dwightguth, @andreistefanescu, please review

File: java-backend/src/main/java/org/kframework/backend/java/kil/Definition.java [hunk 1]
```diff
@@ -149,14 +153,16 @@ public Definition(Context context, KExceptionManager kem, IndexingTable.Data ind
                 if (p.containsAttribute("binder")) {
                     attributes.add(new Attribute<>(
                             Attribute.Key.get(
-                                    new TypeToken<Multimap<Integer, Integer>>() {},
+                                    new TypeToken<Multimap<Integer, Integer>>() {
```
Comment (dwightguth) [inline java-backend/src/main/java/org/kframework/backend/java/kil/Definition.java]: please undo this. This is a really non-nice kind of autoformat.

File: java-backend/src/main/java/org/kframework/backend/java/kil/KItem.java [hunk 2]
```diff
@@ -81,7 +82,7 @@ public static KItem of(Term kLabel, Term kList, TermContext termContext, Source
         return new KItem(kLabel, kList, termContext, termContext.global().kItemOps.tool, source, location);
     }
 
-    KItem(Term kLabel, Term kList, Sort sort, boolean isExactSort) {
+    public KItem(Term kLabel, Term kList, Sort sort, boolean isExactSort) {
```
Comment (dwightguth) [inline java-backend/src/main/java/org/kframework/backend/java/kil/KItem.java:85]: you should be using KItem.of to construct KItems, I don't see why this needs to be public.

**Comment by dwightguth** [comment]
It looks a lot like you did a string search/replace on "Attributes" and replaced it with Att, and it caught in a bunch of places it probably shouldn't have. Can you go through the diff and find the places where this is happening and revert them?

File: kernel/src/main/java/org/kframework/kore/convertors/KOREtoKIL.java [hunk 3]
```diff
@@ -223,15 +222,16 @@ private void generateUserLists() {
...
                 .forEach(
                         sentence -> {
-                            SyntaxProduction syntaxProduction = (SyntaxProduction) sentence;
+                            org.kframework.definition.Production syntaxProduction = (org.kframework.definition.Production) sentence;
                             List<org.kframework.kil.ProductionItem> kilProdItems = new ArrayList<>();
                             for (ProductionItem it : scala.collection.JavaConversions
                                     .seqAsJavaList(syntaxProduction.items())) {
                                 kilProdItems.add(convertProdItem(it));
                             }
+                            System.out.println("sentence = " + sentence);
```
Comment (dwightguth) [inline kernel/src/main/java/org/kframework/kore/convertors/KOREtoKIL.java]: is this a stray debugging relic?

File: kernel/src/main/java/org/kframework/krun/KRunOptions.java [hunk 4]
```diff
@@ -148,7 +148,7 @@ public boolean io() {
     public ColorOptions color = new ColorOptions();
 
     @Parameter(names={"--output", "-o"}, converter=OutputModeConverter.class,
-            description="How to display krun results. <mode> is either [pretty|smart|compatible|kore|raw|binary|none].")
+            description="How to display krun results. <mode> is either [pretty|smart|compatible|koreimplementation|raw|binary|none].")
```
Comment (dwightguth) [inline kernel/src/main/java/org/kframework/krun/KRunOptions.java]: this looks like another search/replace gone wrong?

File: kernel/src/test/java/org/kframework/parser/concrete2kore/TreeCleanerVisitorTest.java [hunk 5]
```diff
@@ -58,6 +56,9 @@ public void testKList() throws Exception {
         assertCleanup(Ambiguity.apply(KList.apply(foo)), foo);
     }
 
+
+//        amb([amb([NOKLABEL(amb([#emptyKRequireList()]),amb([#KModuleList(amb([#KModule(amb([#token(KModuleName,"FOO ")]),amb([#emptyKImportList()]),amb([#emptyKSentenceList()]))]),amb([#emptyKModuleList()]))]))])])
```
Comment (dwightguth) [inline kernel/src/test/java/org/kframework/parser/concrete2kore/TreeCleanerVisitorTest.java]: can I get this line broken?

File: kore/src/main/scala/org/kframework/kore/ADT.scala [hunk 6]
```diff
@@ -0,0 +1,50 @@
+package org.kframework.kore
+
+import org.kframework.kore
+import org.kframework.attributes._
+import collection.JavaConverters._
+
+object ADT {
```
Comment (dwightguth) [inline kore/src/main/scala/org/kframework/kore/ADT.scala:14]: can you add a comment here explaining the acronym?

File: kore/src/main/scala/org/kframework/kore/Unapply.scala [hunk 7]
```diff
@@ -0,0 +1,45 @@
+package org.kframework.kore
+
+import org.kframework.kore
+import collection.JavaConverters._
+
+object Unapply {
```
Comment (dwightguth) [inline kore/src/main/scala/org/kframework/kore/Unapply.scala:10]: comment explaining what this is?

File: kore/src/main/scala/org/kframework/kore/Unparse.scala [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+import Unapply._
+import org.apache.commons.lang3.StringEscapeUtils
+
+/**
+ * Pretty prints inner KORE structures to labeled form.
+ */
+
+object Unparse extends {
+  def apply(k: K): String = k match {
+    case KToken(sort, s) => "#token(" + sort + ",\"" + StringEscapeUtils.escapeJava(s) + "\")"
```
Comment (dwightguth) [inline kore/src/main/scala/org/kframework/kore/Unparse.scala:12]: please add a TODO here for Radu to fix this string syntax.

**Comment by cos** [comment]
this pull request is on hold until @andreistefanescu's changes are remerged (they have been reverted due to some performance issues)

in the meantime, @radumereuta, @bmmoore, please keep in sync with `kore` branch for anything involving kore.

**Comment by cos** [comment]
Jenkings, test this please.

**Comment by cos** [comment]
Jenkins, test this please.

**Comment by cos** [comment]
@dwightguth, I see Andrei's changes have been merged. The pull request still passes the tests (by the way, does Jenkins test this branch, or the merged version?). Are you ok with merging this?

@andreistefanescu, do you want to have a look beforehand?

**Comment by dwightguth** [comment]
It tests the merged version if there are no conflicts, but you haven't resolved the remaining conflicts in the diff, so it hasn't tested the merge yet.

**Comment by cos** [comment]
tests passing. is everybody ok with the merge?

**Comment by radumereuta** [comment]
I haven't looked at the changes, but go ahead. We'll fix any issues on the fly.

**Comment by andreistefanescu** [comment]
As far as I am concerned, you can merge.

**Pull request merged**